### ISSUE TITLE
Transacting collections as the value of a unique, cardinality-many attribute throws a comparison exception

### ISSUE THREAD

**Issue by cjsauer** [tonsky/datascript#320]
Repro adapted from https://github.com/tonsky/datascript/issues/117#issuecomment-141414278

```
(-> (ds/empty-db {:contacts/tag {:db/cardinality :db.cardinality/many
                                 :db/unique      :db.unique/identity}})
    (ds/db-with  [{:db/id 123 :contacts/tag #{:a :b}}])
    (ds/db-with  [{:db/id 123 :contacts/tag #{:c}}])
    (ds/datoms :eavt))


#error {
 :cause "clojure.lang.PersistentHashSet cannot be cast to clojure.lang.Keyword"
 :via
 [{:type java.lang.ClassCastException
   :message "clojure.lang.PersistentHashSet cannot be cast to clojure.lang.Keyword"
   :at [clojure.lang.Keyword compareTo "Keyword.java" 114]}]
 :trace
 [[clojure.lang.Keyword compareTo "Keyword.java" 114]
  [clojure.lang.Util compare "Util.java" 153]
  [datascript.db$cmp invokeStatic "db.cljc" 304]
  [datascript.db$cmp invoke "db.cljc" 301]
  [clojure.lang.AFn applyToHelper "AFn.java" 156]
  [clojure.lang.AFn applyTo "AFn.java" 144]
  [clojure.lang.Var applyTo "Var.java" 705]
  [clojure.core$apply invokeStatic "core.clj" 665]
  [clojure.core$apply invoke "core.clj" 660]
  [cider.nrepl.middleware.debug$apply_instrumented_maybe invokeStatic "debug.clj" 514]
  [cider.nrepl.middleware.debug$apply_instrumented_maybe invoke "debug.clj" 510]
  [datascript.db$eval70053$cmp_datoms_avet__70054 invoke "NO_SOURCE_FILE" 325]
  [clojure.lang.AFunction compare "AFunction.java" 51]
  [me.tonsky.persistent_sorted_set.Leaf searchFirst "Leaf.java" 54]
  [me.tonsky.persistent_sorted_set.PersistentSortedSet slice "PersistentSortedSet.java" 70]
  [me.tonsky.persistent_sorted_set.PersistentSortedSet slice "PersistentSortedSet.java" 49]
  [me.tonsky.persistent_sorted_set$slice invokeStatic "persistent_sorted_set.clj" 30]
  [me.tonsky.persistent_sorted_set$slice invoke "persistent_sorted_set.clj" 24]
  [datascript.db.DB _datoms "db.cljc" 414]
  [datascript.db$upsert_eid$fn__60758 invoke "db.cljc" 999]
  [clojure.lang.PersistentArrayMap kvreduce "PersistentArrayMap.java" 377]
  [clojure.core$fn__8422 invokeStatic "core.clj" 6845]
  [clojure.core$fn__8422 invoke "core.clj" 6830]
  [clojure.core.protocols$fn__8152$G__8147__8161 invoke "protocols.clj" 175]
  [clojure.core$reduce_kv invokeStatic "core.clj" 6856]
  [clojure.core$reduce_kv invoke "core.clj" 6847]
  [datascript.db$upsert_eid invokeStatic "db.cljc" 996]
  [datascript.db$upsert_eid invoke "db.cljc" 993]
  [datascript.db$transact_tx_data invokeStatic "db.cljc" 1161]
  [datascript.db$transact_tx_data invoke "db.cljc" 1124]
  [datascript.core$with invokeStatic "core.cljc" 228]
  [datascript.core$with invoke "core.cljc" 221]
  [datascript.core$with invokeStatic "core.cljc" 223]
  [datascript.core$with invoke "core.cljc" 221]
  [datascript.core$db_with invokeStatic "core.cljc" 240]
  [datascript.core$db_with invoke "core.cljc" 236]
  [splitpea.server$eval71809 invokeStatic "NO_SOURCE_FILE" 100]
  [splitpea.server$eval71809 invoke "NO_SOURCE_FILE" 97]
  [clojure.lang.Compiler eval "Compiler.java" 7176]
  [clojure.lang.Compiler eval "Compiler.java" 7131]
  [clojure.core$eval invokeStatic "core.clj" 3214]
  [clojure.core$eval invoke "core.clj" 3210]
  [clojure.main$repl$read_eval_print__9068$fn__9071 invoke "main.clj" 414]
  [clojure.main$repl$read_eval_print__9068 invoke "main.clj" 414]
  [clojure.main$repl$fn__9077 invoke "main.clj" 435]
  [clojure.main$repl invokeStatic "main.clj" 435]
  [clojure.main$repl doInvoke "main.clj" 345]
  [clojure.lang.RestFn applyTo "RestFn.java" 137]
  [clojure.core$apply invokeStatic "core.clj" 665]
  [clojure.core$apply invoke "core.clj" 660]
  [refactor_nrepl.ns.slam.hound.regrow$wrap_clojure_repl$fn__61380 doInvoke "regrow.clj" 18]
  [clojure.lang.RestFn invoke "RestFn.java" 1523]
  [nrepl.middleware.interruptible_eval$evaluate invokeStatic "interruptible_eval.clj" 79]
  [nrepl.middleware.interruptible_eval$evaluate invoke "interruptible_eval.clj" 55]
  [nrepl.middleware.interruptible_eval$interruptible_eval$fn__24080$fn__24084 invoke "interruptible_eval.clj" 142]
  [clojure.lang.AFn run "AFn.java" 22]
  [nrepl.middleware.session$session_exec$main_loop__24307$fn__24311 invoke "session.clj" 171]
  [nrepl.middleware.session$session_exec$main_loop__24307 invoke "session.clj" 170]
  [clojure.lang.AFn run "AFn.java" 22]
  [java.lang.Thread run "Thread.java" 748]]}
```

**Comment by cjsauer** [comment]
EDIT: include exception with source maps

**Comment by cjsauer** [comment]
I'm not super familiar with the source, but it seems that datom comparison does not take the schema into account, but compares value-to-value: https://github.com/tonsky/datascript/blob/f18db4e595692603e7c0904b7d0ef96837d586da/src/datascript/db.cljc#L323-L328
Transactions of this nature work in datomic, and so I wonder if they're doing a more advanced comparison that includes the respective attribute cardinality 🤔

**Comment by cjsauer** [comment]
Tracked the exception down to the `-datoms` call that happens in `upsert-eid` in the `db.cljc` file. For cardinality many attributes, when a collection value is passed as `v`, that set slice could potentially start at `(-> v sort first)`. There is likely a better way to do this, but this at least passes the above test:

```
(-> (ds/empty-db {:contacts/tag {:db/cardinality :db.cardinality/many
                                 :db/unique      :db.unique/identity}})
    (ds/db-with  [{:db/id 123 :contacts/tag #{:a :b}}])
    (ds/db-with  [{:db/id 123 :contacts/tag #{:c}}])
    (ds/datoms :eavt))
(#datascript/Datom [123 :contacts/tag :a 536870913 true]
 #datascript/Datom [123 :contacts/tag :b 536870913 true]
 #datascript/Datom [123 :contacts/tag :c 536870914 true])
```

**Comment by tonsky** [comment]
Are you sure they work in Datomic? From https://docs.datomic.com/on-prem/identity.html:

> Uniqueness can be declared on attributes of any value type, including references (:db.type/ref). Only (:db.cardinality/one) attributes can be unique.

Although I’m not sure this limitation has a good reason to exist. At least I don’t see what could be wrong with cardinality many unique attribute

**Comment by tonsky** [comment]
Looking at your PR, I remember now. Upsert would not work because it’s not clear which value to look at. Or we must look at all provided values and make sure they all resolve to the same entity. What’s your use case anyway?

**Comment by cjsauer** [comment]
Interesting. It appears to work correctly in Datomic Cloud:

Assuming this schema:
```
   {:db/ident       :user/email
    :db/unique      :db.unique/identity
    :db/valueType   :db.type/string
    :db/cardinality :db.cardinality/many}
```

The following transactions work fine:
```
(d/transact conn {:tx-data [{:user/email "calvin"}]})
=> {:tx-data [#datom[13194139533323 50 #inst "2019-10-17T13:44:42.951-00:00" 13194139533323 true] #datom[10740029580116059 78 "calvin" 13194139533323 true]]

(d/transact conn {:tx-data [{:user/email ["jenny"]}]})
=> [#datom[13194139533325 50 #inst "2019-10-17T13:47:13.676-00:00" 13194139533325 true] #datom[13453624277467228 78 "jenny" 13194139533325 true]]
```

Pulls work as well:
```
(d/pull db [:user/email] [:user/email "calvin"])
=> #:user{:email ["calvin"]}
```

Schema proof:
```
(d/pull (irope/get-db config) '[*] :user/email)
#:db{:id 78,
     :ident :user/email,
     :valueType #:db{:id 23, :ident :db.type/string},
     :cardinality #:db{:id 36, :ident :db.cardinality/many},
     :unique #:db{:id 38, :ident :db.unique/identity}}
```
I have brought this up in the [#datomic channel](https://clojurians.slack.com/archives/C03RZMDSH/p1571320110249800) of the Clojure Slack, and Marshall is equally puzzled. Getting to the bottom of it...

**Comment by cjsauer** [comment]
@tonsky well, apparently it *can* be done, but you *shouldn't*. I agree tho that this doesn't seem like a technical limitation. The discussion unfortunately didn't lead to a satisfying conclusion (as of yet).

>What’s your use case anyway?
The example above: modeling users as being uniquely identifiable by multiple email addresses (which matches reality in my opinion). It's unfortunate that this semantic isn't fully supported.

**Comment by cjsauer** [comment]
Datomic forum post on the matter: https://forum.datomic.com/t/unique-cardinality-many-attributes-why-not/1208

**Comment by tonsky** [comment]
I think we can support this use-case. But the semantic should be the same as with having multiple unique attributes in a tx on a single entity: they either all not resolve to anything (and in that case we create a new entity) or part of them resolves to the same entity (and the rest will be added and rest of the attributes merged to that entity). Care to provide a PR and tests?

**Comment by cjsauer** [comment]
@tonsky happy to. Here's what I'm thinking https://github.com/tonsky/datascript/pull/321/files

It's a bit rough, and I'm unsure of the performance characteristics of `-datoms`, but this is the general idea (with passing tests). Let me know what you think.

**Comment by tonsky** [comment]
Fixed in 0.18.7
